doc/man/man1/virgilverify.1
.TH "VIRGIL" "1" "August 01, 2015" "Virgil Security CLI (0.3.0)" "BSD General Commands Manual"
.SH NAME
.PP
\f[B]verify\f[] \-\- verify data.
.SH SYNOPSIS
.PP
\f[B]virgil verify\f[] [\-\-in \f[I]file\f[]] [\-\-out \f[I]file\f[]]
\-\-sign \f[I]file\f[] \-\-sign\-owner \f[I]arg\f[]
.SH DESCRIPTION
.PP
Verify data with given user\[aq]s identifier or with it Virgil Public
Key.
.SH OPTIONS
.TP
.B \-i \f[I]file\f[], \-\-in \f[I]file\f[]
Data to be verified.
If omitted stdin is used.
.RS
.RE
.TP
.B \-o \f[I]file\f[], \-\-out \f[I]file\f[]
Verification result: \f[C]success\f[] | \f[C]failure\f[].
If omitted stdout is used.
.RS
.RE
.TP
.B \-s \f[I]file\f[], \-\-sign \f[I]file\f[]
(required) Digest sign.
.RS
.RE
.TP
.B \-r \f[I]arg\f[], \-\-sign\-owner \f[I]arg\f[]
(required) Sign owner, defined in format:
.RS
.IP
.nf
\f[C]
[file|email|phone|domain]:<value>
\f[]
.fi
.PP
where:
.IP \[bu] 2
if \f[C]file\f[], then \f[I]value\f[] \- signers\[aq]s Virgil Public Key
file stored locally;
.IP \[bu] 2
if \f[C]email\f[], then \f[I]value\f[] \- signers\[aq]s email;
.IP \[bu] 2
if \f[C]phone\f[], then \f[I]value\f[] \- signers\[aq]s phone;
.IP \[bu] 2
if \f[C]domain\f[], then \f[I]value\f[] \- signers\[aq]s domain.
.RE
.SH SEE ALSO
.PP
\f[C]virgil(1)\f[], \f[C]virgilsign(1)\f[]
.SH AUTHORS
(c) Virgil Security Inc.


ChangeLog
virgil CLI ChangeLog (Sorted per date)

= Version 0.3.0 released 2015-08-13

Improvements:
  * Add Markdown documentation.

Changes:
  * Replace certificate term with Virgil Public Key term.
  * Remove options to specify sign's output format.
  * Use C++11 standard.

Features:
  * Add command "keyreg" - register user's public key on the Virgil Public Keys service.
  * Add command "keyget" - get user's public key from the Virgil Public Keys service.
  * Add command "confirm" - send confirmation code to the Virgil Public Keys service.
  * Add command "reconfirm" - resend confirmation code to the user for given user's identifier.

= Version 0.2.0 released 2015-05-15

Bugfix:
  * Use standard constructor of class 'ifstream'.
  * Change parameter 'p' long name from 'arg' to 'pwd'.
  * Add required include statement <cstdlib>.

Improvements:
  * Fix misspelling in the help text.
  * Add "happy path" test.
  * Modify 'encrypt' and 'verify' commands to work with Virgil PKI service thru user's identifier.

Changes:
  * Change parameter 'asn1' to 'der'.
  * Change parameters format for 'encrypt' command.

Features:
  * Add 'verify' command.
  * Add 'adduser' command

= Version 0.1.0 released 2015-05-04

Features:
  * Add command "keygen" - generate private key with given parameters.
  * Add command "key2pub" - extract public key from the private key.
  * Add command "pub2cert" - create certificate from the public key and identifiers.
  * Add command "certinfo" - output detailed information about given certificate.
  * Add command "encrypt" - encrypt data for given recipients which can be defined by certificates and by passwords.
  * Add command "decrypt" - decrypt data for given recipient which can be defined by certificate or by password.
  * Add command "sign" - sign data with private key.
  * Add command "verify" - verify data with certificate.


src/commands/virgilverify.cxx
/**
 * Copyright (C) 2015 Virgil Security Inc.
 *
 * Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     (1) Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *     (2) Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *     (3) Neither the name of the copyright holder nor the names of its
 *     contributors may be used to endorse or promote products derived from
 *     this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <fstream>
#include <iostream>
#include <iterator>
#include <stdexcept>
#include <string>

#include <tclap/CmdLine.h>

#include <virgil/crypto/VirgilByteArray.h>
#include <virgil/crypto/VirgilStreamSigner.h>
#include <virgil/crypto/stream/VirgilStreamDataSource.h>

#include <virgil/sdk/keys/model/PublicKey.h>

#include <cli/version.h>
#include <cli/pair.h>
#include <cli/util.h>

using virgil::crypto::VirgilByteArray;
using virgil::crypto::VirgilStreamSigner;
using virgil::crypto::stream::VirgilStreamDataSource;

using virgil::sdk::keys::model::PublicKey;


#ifdef SPLIT_CLI
#define MAIN main
#else
#define MAIN verify_main
#endif

int MAIN(int argc, char **argv)
{
    try {
        // Parse arguments.
        TCLAP::CmdLine cmd("Verify data with given user's identifier or with it Virgil Public Key.", ' ',
                virgil::cli_version());

        TCLAP::ValueArg<std::string> inArg("i", "in", "Data to be verified. If omitted stdin is used.",
                false, "", "file");

        TCLAP::ValueArg<std::string> outArg("o", "out",
                "Verification result: success | failure. If omitted stdout is used.",
                false, "", "file");

        TCLAP::ValueArg<std::string> signArg("s", "sign", "Digest sign.",
                true, "", "file");

        TCLAP::ValueArg<std::string> signOwnerArg("r", "sign-owner",
                "Sign owner, defined in format:"
                "[file|email|phone|domain]:<value>\n"
                "where:\n"
                "\t* if file, then <value> - signers's Virgil Public Key file stored locally;\n"
                "\t* if email, then <value> - signers's email;\n"
                "\t* if phone, then <value> - signers's phone;\n"
                "\t* if domain, then <value> - signers's domain.\n",
                true, "", "arg");


        cmd.add(signOwnerArg);
        cmd.add(signArg);
        cmd.add(outArg);
        cmd.add(inArg);

        cmd.parse(argc, argv);

        // Prepare input
        std::istream* inStream;
        std::ifstream inFile;
        if (inArg.getValue().empty() || inArg.getValue() == "-") {
            inStream = &std::cin;
        } else {
            inFile.open(inArg.getValue(), std::ios::in | std::ios::binary);
            if (!inFile) {
                throw std::invalid_argument("can not read file: " + inArg.getValue());
            }
            inStream = &inFile;
        }

        // Verify data
        VirgilStreamDataSource dataSource(*inStream);

        // Read sign
        std::ifstream signFile(signArg.getValue(), std::ios::in | std::ios::binary);
        if (!signFile) {
            throw std::invalid_argument("can not read file: " + signArg.getValue());
        }

        VirgilByteArray sign((std::istreambuf_iterator<char>(signFile)), std::istreambuf_iterator<char>());

        const std::pair<std::string, std::string> recipient = virgil::cli::parse_pair(signOwnerArg.getValue());
        const std::string recipientIdType = recipient.first;
        const std::string recipientId = recipient.second;

        PublicKey publicKey;
        if(recipientIdType == "file") {
            // Read Virgil Public Key
            std::ifstream virgilPublicKeyFile(recipientId, std::ios::in | std::ios::binary);
            if (!virgilPublicKeyFile) {
                throw std::invalid_argument("can not read Virgil Public Key: " + recipientId);
            }
            publicKey = virgil::cli::read_virgil_public_key(virgilPublicKeyFile);

        } else {
            // Get Virgil Public Key from the service
           publicKey = virgil::cli::get_virgil_public_key(recipientId, recipientIdType);
        }

        // Create signer
        VirgilStreamSigner signer;
        bool verified = signer.verify(dataSource, sign, publicKey.key());
        if (verified) {
            virgil::cli::write_bytes(outArg.getValue(), "success");
            return EXIT_SUCCESS;
        } else {
            virgil::cli::write_bytes(outArg.getValue(), "failure");
            return EXIT_FAILURE;
        }
    } catch (TCLAP::ArgException& exception) {
        std::cerr << "Error: " << exception.error() << " for arg " << exception.argId() << std::endl;
        return EXIT_FAILURE;
    } catch (std::exception& exception) {
        std::cerr << "Error: " << exception.what() << std::endl;
        return EXIT_FAILURE;
    }
}
